server/block/sugar_cane.go

```diff
@@ -0,0 +1,101 @@
+package block
+
+import (
+	"github.com/df-mc/dragonfly/server/block/cube"
+	"github.com/df-mc/dragonfly/server/item"
+	"github.com/df-mc/dragonfly/server/world"
+	"github.com/df-mc/dragonfly/server/world/particle"
+	"github.com/go-gl/mathgl/mgl64"
+	"math/rand"
+)
+
+// SugarCane is a plant block that generates naturally near water.
+type SugarCane struct {
+	empty
+	transparent
+
+	// Age is the growth state of sugar cane. Values range from 0 to 15.
+	Age int
+}
+
+// UseOnBlock ensures the placement of the block is OK.
+func (c SugarCane) UseOnBlock(pos cube.Pos, face cube.Face, _ mgl64.Vec3, w *world.World, user item.User, ctx *item.UseContext) (used bool) {
+	pos, _, used = firstReplaceable(w, pos, face, c)
+	if !used {
+		return false
+	}
+	if !c.CanGrowHere(pos, w, 3) {
```

Comment:
This isn't quite right as you can continue stacking sugarcane through placement. In fact, sugar cane in vanilla Minecraft can naturally generate 4 blocks tall!

Reply:
I have removed the height checking to allow higher placements.